### DISCUSSION TITLE
Question on RSA context of CA certificate clean up by x509_free

### DISCUSSION THREAD

**Discussion by bhatja1** [openssl/openssl#24457] [Q&A] [answered]
Hi All,

During TLS handshake _method_mod_n gets allocated in RSA context of CA certificate in rsa_ossl_public_decrypt function. After the handshake we are cleaning up SSL and SSL context. x509_free is called for the certificate clean up. All allocated resources are cleaned up except the **_method_mod_n**. This is resulting in a leak.  Is this expected behaviour?  We are using OpenSSL 1.1.1w.

To give more details 

In rsa_ossl_public_decrypt function in file name rsa_ossl.c.  **"rsa structure"** from the installed CA that used to validate server certificate.  **_method_mod_n**  parameter in **"rsa structure"** is allocated as below.

```` C
   if (rsa->flags & RSA_FLAG_CACHE_PUBLIC)
   {
        if (!BN_MONT_CTX_set_locked(&rsa->_method_mod_n, rsa->lock,
                                    rsa->n, ctx))
	{
            goto err;
	}
    }
````

But we saw that as part of clean up X509_STORE_CTX_free is called, we see that above parameter "_method_mod_n " is not getting freed. Is there any additional calls needed from application? Is is it addressed in higher versions like 3.x?

Regards,
Jayalakshmi

**Comment by mattcaswell** [comment, marked as answer]
The `RSA_FLAG_CACHE_PUBLIC` flag (on by default) instructs OpenSSL to cache this value in the `RSA` object. It gets freed when the `RSA` object gets freed - not when the certificate is freed:

```` C
static int rsa_ossl_finish(RSA *rsa)
{
    int i;
    RSA_PRIME_INFO *pinfo;

    BN_MONT_CTX_free(rsa->_method_mod_n);
    BN_MONT_CTX_free(rsa->_method_mod_p);
    BN_MONT_CTX_free(rsa->_method_mod_q);
    for (i = 0; i < sk_RSA_PRIME_INFO_num(rsa->prime_infos); i++) {
        pinfo = sk_RSA_PRIME_INFO_value(rsa->prime_infos, i);
        BN_MONT_CTX_free(pinfo->m);
    }
    return 1;
}

**Comment by bhatja1** [comment]
Hi @mattcaswell 

Thanks alot for the response.  I could track it to our application issue.  We were wrongly handling the certificates in cert store

Regards,
Jayalakshmi
